File: sdk/Php/CPaySigning.php
<?php

final class CPaySigning
{
    public static function sha256Hex(string $value): string
    {
        return hash('sha256', $value);
    }

    public static function canonicalQuery(array $query = []): string
    {
        if (!$query) {
            return '';
        }
        ksort($query);
        $parts = [];
        foreach ($query as $key => $value) {
            if ($value === null) {
                continue;
            }
            $parts[] = rawurlencode((string) $key) . '=' . rawurlencode((string) $value);
        }
        return implode('&', $parts);
    }

    public static function canonicalString(
        string $method,
        string $path,
        array $query,
        string $timestamp,
        string $nonce,
        string $body = ''
    ): string {
        return implode("\n", [
            strtoupper($method ?: 'GET'),
            $path ?: '/',
            self::canonicalQuery($query),
            $timestamp,
            $nonce,
            self::sha256Hex($body),
        ]);
    }

    public static function signRequest(
        string $merchantNumber,
        string $privateKeyPem,
        string $method,
        string $path,
        array $query = [],
        string $body = '',
        ?string $timestamp = null,
        ?string $nonce = null,
        ?string $idempotencyKey = null
    ): array {
        if ($merchantNumber === '') {
            throw new InvalidArgumentException('merchantNumber is required');
        }
        if ($privateKeyPem === '') {
            throw new InvalidArgumentException('privateKeyPem is required');
        }
        $timestamp = $timestamp ?: gmdate('Y-m-d\TH:i:s\Z');
        $nonce = $nonce ?: bin2hex(random_bytes(16));
        $idempotencyKey = $idempotencyKey ?: bin2hex(random_bytes(16));
        $canonical = self::canonicalString($method, $path, $query, $timestamp, $nonce, $body);
        $ok = openssl_sign($canonical, $signature, $privateKeyPem, OPENSSL_ALGO_SHA256);
        if (!$ok) {
            throw new RuntimeException('Unable to sign CPay request');
        }

        return [
            'X-CPay-Merchant-Number' => $merchantNumber,
            'X-CPay-Signature-Version' => 'v2',
            'X-CPay-Timestamp' => $timestamp,
            'X-CPay-Nonce' => $nonce,
            'X-CPay-Signature' => base64_encode($signature),
            'X-CPay-Idempotency-Key' => $idempotencyKey,
        ];
    }
}

File: Sdk/Python/cpay_signing.py
import base64
import hashlib
import uuid
from datetime import datetime, timezone
from urllib.parse import quote

from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import padding


def sha256_hex(value: str) -> str:
    return hashlib.sha256((value or "").encode("utf-8")).hexdigest()


def canonical_query(query=None) -> str:
    if not query:
        return ""
    items = query.items() if hasattr(query, "items") else query
    return "&".join(
        f"{quote(str(key), safe='')}={quote(str(value), safe='')}"
        for key, value in sorted(items, key=lambda item: str(item[0]))
        if value is not None
    )


def canonical_string(method: str, path: str, query, timestamp: str, nonce: str, body: str = "") -> str:
    return "\n".join([
        (method or "GET").upper(),
        path or "/",
        canonical_query(query),
        timestamp,
        nonce,
        sha256_hex(body or ""),
    ])


def sign_request(
    merchant_number: str,
    private_key_pem: str,
    method: str,
    path: str,
    query=None,
    body: str = "",
    timestamp: str = None,
    nonce: str = None,
    idempotency_key: str = None,
):
    if not merchant_number:
        raise ValueError("merchant_number is required")
    if not private_key_pem:
        raise ValueError("private_key_pem is required")

    timestamp = timestamp or datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")
    nonce = nonce or str(uuid.uuid4())
    idempotency_key = idempotency_key or str(uuid.uuid4())
    canonical = canonical_string(method, path, query, timestamp, nonce, body)
    private_key = serialization.load_pem_private_key(private_key_pem.encode("utf-8"), password=None)
    signature = private_key.sign(canonical.encode("utf-8"), padding.PKCS1v15(), hashes.SHA256())

    return {
        "X-CPay-Merchant-Number": merchant_number,
        "X-CPay-Signature-Version": "v2",
        "X-CPay-Timestamp": timestamp,
        "X-CPay-Nonce": nonce,
        "X-CPay-Signature": base64.b64encode(signature).decode("ascii"),
        "X-CPay-Idempotency-Key": idempotency_key,
    }
